File: packages/zod/src/policy.ts
import { z } from "zod";

export const ZPolicyStatus = z.enum([
  "pending",
  "ingesting",
  "active",
  "failed",
  "archived",
]);
export type PolicyStatus = z.infer<typeof ZPolicyStatus>;

export const ZPolicy = z.object({
  id: z.string().uuid(),
  name: z.string(),
  gcsPath: z.string(),
  version: z.string(),
  status: ZPolicyStatus,
  chunkCount: z.number().int(),
  uploadedBy: z.string(),
  createdAt: z.string(),
  updatedAt: z.string(),
});
export type Policy = z.infer<typeof ZPolicy>;

export const ZUploadPolicyResponse = ZPolicy;
export type UploadPolicyResponse = Policy;

export const ZListPoliciesResponse = z.array(ZPolicy);
export type ListPoliciesResponse = z.infer<typeof ZListPoliciesResponse>;

File: apps/backend/internal/model/policy.go
package model

import (
	"encoding/json"

	"github.com/google/uuid"
)

type PolicyStatus string

const (
	PolicyStatusPending   PolicyStatus = "pending"
	PolicyStatusIngesting PolicyStatus = "ingesting"
	PolicyStatusActive    PolicyStatus = "active"
	PolicyStatusFailed    PolicyStatus = "failed"
	PolicyStatusArchived  PolicyStatus = "archived"
)

type Policy struct {
	Base
	Name       string       `json:"name"       db:"name"`
	GCSPath    string       `json:"gcsPath"    db:"gcs_path"`
	Version    string       `json:"version"    db:"version"`
	Status     PolicyStatus `json:"status"     db:"status"`
	ChunkCount int          `json:"chunkCount" db:"chunk_count"`
	UploadedBy string       `json:"uploadedBy" db:"uploaded_by"`
	OrgID      string       `json:"orgId"      db:"org_id"`
}

type PolicyChunk struct {
	Base
	Policy     uuid.UUID `db:"policy_id"`
	ChunkText  string    `db:"chunk_text"`
	Category   string    `db:"category"`
	PageNum    int       `db:"page_num"`
	ChunkIndex int       `db:"chunk_index"`
}

type PolicyChunkInsert struct {
	PolicyID   uuid.UUID
	ChunkText  string
	Embedding  []float32
	Category   string
	PageNum    int
	ChunkIndex int
}

type RetrievedChunk struct {
	ChunkText string  `json:"chunk_text"`
	Category  string  `json:"category"`
	PageNum   int     `json:"page_num"`
	Score     float64 `json:"score"`
}

func MarshalRetrievedChunks(chunks []RetrievedChunk) ([]byte, error) {
	return json.Marshal(chunks)
}

func UnmarshalRetrievedChunks(raw []byte) ([]RetrievedChunk, error) {
	if len(raw) == 0 {
		return nil, nil
	}
	var chunks []RetrievedChunk
	if err := json.Unmarshal(raw, &chunks); err != nil {
		return nil, err
	}
	return chunks, nil
}
